synapse/storage/databases/main/presence.py

```diff
@@ -157,5 +157,64 @@ async def get_presence_for_users(self, user_ids):
 
         return {row["user_id"]: UserPresenceState(**row) for row in rows}
 
+    async def get_presence_for_all_users(
+        self,
+        include_offline: bool = True,
+    ) -> Dict[str, UserPresenceState]:
+        """Retrieve the current presence state for all users.
+
+        Note that the presence_stream table is culled frequently, so it should only
+        contain the latest presence state for each user.
+
+        Args:
+            include_offline: Whether to include offline presence states
+
+        Returns:
+            A dict of user IDs to their current UserPresenceState.
+        """
+        users_to_state = {}
+
+        exclude_keyvalues = {}
+        if not include_offline:
+            # Exclude offline presence state
+            exclude_keyvalues = {"state": "offline"}
+
+        # This may be a very heavy database query.
+        # We paginate in order to not block a database connection.
+        limit = 100
```

Comment:
100 seems kind of low, I think we usually batch by 1000?

Reply:
The only other usage of this function I see uses a limit of 100:

https://github.com/matrix-org/synapse/blob/0a00b7ff14890987f09112a2ae696c61001e6cf1/synapse/storage/databases/main/stats.py#L297-L320

But intuitively 1000 is probably also fine, given the overhead of spinning up another query?

Reply:
Fair enough. We can tweak it if we need to.